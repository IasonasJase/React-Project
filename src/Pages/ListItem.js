import React, { useState } from "react";
import { useDispatch } from "react-redux";
import { removeTodoAction, updateTodoAction, toggleTodoAction } from "./todoActions";

const ListItem = (props) => {
  const [input, setInput] = useState(props.item.title);
  const [hidden, setHidden] = useState(false);
  const dispatch = useDispatch();
  const { item } = props;

  return (
    <li data-testid={item.id} className={!item.completed ? "todo-collection__item " : "todo-collection__item fade"}>
      <span className={!hidden ? "todo-collection__item__title" : "todo-collection__item__title hidden"}>
        {item.title}
      </span>
      <input
        value={input}
        id={item.id}
        className={!hidden ? "hidden input input--todo" : "input input--todo"}
        onChange={(e) => setInput(e.target.value)}
      ></input>
      <button
        type="submit"
        onClick={() => {
          dispatch(updateTodoAction(input, item.id));
          setHidden(false);
        }}
        className={!hidden ? "button button--todo button--save hidden" : "button button--todo button--save"}
      >
        Save
      </button>
      <button
        onClick={() => setHidden(true)}
        className={!hidden ? "button button--todo button--edit " : "button button--todo button--edit hidden"}
      >
        Edit
      </button>
      <button
        data-testid={`Delete${item.id}`}
        onClick={() => dispatch(removeTodoAction(item.id))}
        className="button button--todo button--delete"
      >
        Delete
      </button>
      <button
        onClick={() => {
          dispatch(toggleTodoAction(item.id));
        }}
        className="button button--todo button--delete"
      >
        Complete
      </button>
    </li>
  );
};

export default ListItem;
