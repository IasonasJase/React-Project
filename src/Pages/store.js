import { createStore } from "redux";
import { combineReducers, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import { composeWithDevTools } from "redux-devtools-extension";
import { todoReducer } from "./todoReducers";
import { saveState, loadState } from "./localstorage";
import throttle from "lodash.throttle";

const reducer = combineReducers({
  todos: todoReducer,
});

const middleware = [thunk];

const store = createStore(reducer, loadState(), composeWithDevTools(applyMiddleware(...middleware)));

store.subscribe(throttle(() => saveState(store.getState()), 1000));
export default store;
