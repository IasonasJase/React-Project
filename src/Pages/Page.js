import React, { useState } from "react";
import "./Pagecss.css";
import { useDispatch, useSelector } from "react-redux";
import ListItem from "./ListItem";
import { addTodoAction, fetchingData, clearTodoAction } from "./todoActions";

const DoPage = () => {
  const [title, setTitle] = useState("");
  const [shake, setShake] = useState(false);
  const dispatch = useDispatch();
  const todos = useSelector((state) => state.todos);

  const handleChange = (event) => setTitle(event.target.value);

  const addToDo = (e) => {
    e.preventDefault();
    if (title.trim().length > 0) {
      dispatch(addTodoAction(title));
      setShake(false);
    } else {
      setShake(true);
      setTimeout(() => {
        setShake(false);
      }, 250);
    }
    setTitle("");
  };

  const handleFetch = () => {
    dispatch(fetchingData());
  };

  const handleClear = () => {
    dispatch(clearTodoAction());
  };

  return (
    <main className="wrapper">
      <nav className="navbar">
        <ul className="nav">
          <li className="nav__item">TODO</li>
        </ul>
      </nav>
      <form className={!shake ? "todo-form" : "todo-form shake-horizontal"} onSubmit={addToDo}>
        <div className="input-field">
          <input
            data-testid="input"
            type="text"
            id="input--add"
            className="input"
            placeholder="e.g. Go to the gym"
            onChange={handleChange}
            value={title}
          />
          <label htmlFor="input--add"></label>
        </div>
        <button data-testid="Add" type="submit" className="button button--add">
          Add
        </button>
      </form>
      <div className="nav nav__item">
        <button data-testid="Fetch" className="button--todo" onClick={handleFetch}>
          Fetch
        </button>
      </div>
      <button className="button--todo" onClick={handleClear}>
        Clear
      </button>
      <ul className="todo-collection" data-testid="testidforli">
        {todos && todos.map((item) => <ListItem key={item.id} item={item} />)}
      </ul>
    </main>
  );
};

export default DoPage;
