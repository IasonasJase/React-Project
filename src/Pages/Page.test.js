import Page from "./Page";
import ListItem from "./ListItem";
import React from "react";
import { createStore } from "redux";
import { render, fireEvent, cleanup, screen, within, getByText } from "@testing-library/react";
import { Provider } from "react-redux";
import { ADD_TODO, DELETE_TODO } from "./todoActions";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";
import { todoReducer } from "./todoReducers";
import pretty from "pretty";
import store from "./store";
import { combineReducers } from "redux";

function renderWithRedux(component, inititalState) {
  const reducer = combineReducers({
    todos: todoReducer,
  });
  const store = createStore(reducer, inititalState);
  return {
    ...render(<Provider store={store}>{component}</Provider>),
  };
}
describe("Page Component", () => {
  it("renders correctly", () => {
    const { container } = renderWithRedux(<Page />);
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  test("renders Page with redux", () => {
    renderWithRedux(<Page />);
    const element = screen.getByTestId("testidforli");
    expect(element).toHaveTextContent("");
  });

  test("change value of input works", () => {
    renderWithRedux(<Page />);
    const inputEl = screen.getByTestId("input");
    fireEvent.change(inputEl, {
      target: {
        value: "football",
      },
    });
    expect(inputEl.value).toBe("football");
  });

  test("add a todo to the list", () => {
    renderWithRedux(<Page />);

    renderWithRedux(<ListItem item={{ id: 1, title: "hello-world-j", completed: false }} />);
    const liEl = screen.getByTestId(1);

    expect(liEl).toBeInTheDocument();
    expect(liEl).toBeVisible();
    expect(liEl).toHaveTextContent("hello-world-j");
  });

  test("a todo's title from the list", () => {
    renderWithRedux(<Page />);

    renderWithRedux(<ListItem item={{ id: 1, title: "hello-world", completed: false }} />);
    const liEl = screen.getByTestId(1);

    expect(liEl).toHaveTextContent("hello-world");
  });

  test("remove a todo from the list", () => {
    renderWithRedux(<Page />, { todos: [{ id: 1, title: "dada", completed: false }] });
    expect(screen.getByText("dada")).toBeInTheDocument();
    const inputEl = screen.getByTestId("input");
    fireEvent.change(inputEl, {
      target: {
        value: "football",
      },
    });

    const addButton = screen.getByText("Add");
    fireEvent.click(addButton);
    const liEl = screen.getByText("football");
    fireEvent.click(screen.getAllByText("Delete")[1]);
    expect(liEl).not.toBeInTheDocument();
  });

  test("should reset the input", () => {
    renderWithRedux(<Page />);

    const inputValue = screen.getByPlaceholderText("e.g. Go to the gym");
    const buttonEl = screen.getByTestId("Add");
    fireEvent.click(buttonEl);
    expect(inputValue.value).toMatch("");
  });
});
