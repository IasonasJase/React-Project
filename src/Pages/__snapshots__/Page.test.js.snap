// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Component renders correctly 1`] = `
"<main class=\\"wrapper\\">
  <nav class=\\"navbar\\">
    <ul class=\\"nav\\">
      <li class=\\"nav__item\\">TODO</li>
    </ul>
  </nav>
  <form class=\\"todo-form\\">
    <div class=\\"input-field\\"><input data-testid=\\"input\\" type=\\"text\\" id=\\"input--add\\" class=\\"input\\" placeholder=\\"e.g. Go to the gym\\" value=\\"\\"><label for=\\"input--add\\"></label></div><button data-testid=\\"Add\\" type=\\"submit\\" class=\\"button button--add\\">Add</button>
  </form>
  <div class=\\"nav nav__item\\"><button data-testid=\\"Fetch\\" class=\\"button--todo\\">Fetch</button></div><button class=\\"button--todo\\">Clear</button>
  <ul class=\\"todo-collection\\" data-testid=\\"testidforli\\"></ul>
</main>"
`;
